frame/contracts/src/gas_weight_conv.rs

```diff
@@ -0,0 +1,195 @@
+	transaction_validity::{
+		InvalidTransaction, TransactionValidity, TransactionValidityError, ValidTransaction,
+	},
+};
+use sp_std::{convert::TryFrom, fmt, marker::PhantomData, prelude::*};
+
+/// Data which is collected during the pre-dispatch phase and needed at the post_dispatch phase.
+#[cfg_attr(test, derive(Debug))]
+pub struct GasCallEnvelope<AccountId, NegativeImbalance> {
+	/// The account id who should receive the refund from the gas leftovers.
+	transactor: AccountId,
+	/// The negative imbalance obtained by withdrawing the value up to the requested gas limit.
+	imbalance: NegativeImbalance,
+}
+
+/// A `SignedExtension` required to properly handle gas limits requested by contract executing
+/// transactions.
+#[derive(codec::Encode, codec::Decode, Clone, Eq, PartialEq)]
+pub struct GasWeightConversion<T: Trait + Send + Sync>(PhantomData<T>);
+
+impl<T: Trait + Send + Sync> GasWeightConversion<T> {
+	pub fn new() -> Self {
+		Self(PhantomData)
+	}
+
+	/// Perform pre-dispatch checks for the given call and origin.
+	fn perform_pre_dispatch_checks(
+		who: &T::AccountId,
+		call: &<T as Trait>::Call,
+	) -> Result<
+		Option<GasCallEnvelope<T::AccountId, NegativeImbalanceOf<T>>>,
+		TransactionValidityError,
+	> {
+		// This signed extension only deals with this module's `Call`.
+		let call = match call.is_sub_type() {
+			Some(call) => call,
+			None => return Ok(None),
+		};
+
+		match *call {
+			Call::claim_surcharge(_, _) | Call::update_schedule(_) | Call::put_code(_) => Ok(None),
+			Call::call(_, _, gas_limit, _) | Call::instantiate(_, gas_limit, _, _) => {
+				// Compute how much block weight this transaction can take at its limit, i.e.
+				// if this transaction depleted all provided gas to zero.
+				let gas_weight_limit = Weight::try_from(gas_limit)
+					.map_err(|_| InvalidTransaction::ExhaustsResources)?;
+				let weight_available = <frame_system::Module<T>>::remaining_weight().into();
+				if gas_weight_limit > weight_available {
+					// We discard the transaction if the requested limit exceeds the available
+					// amount of weight in the current block.
+					//
+					// Note that this transaction is left out only from the current block and txq
+					// might attempt to include this transaction again.
+					return Err(InvalidTransaction::ExhaustsResources.into());
+				}
+
+				// Compute the fee corresponding for the given gas_weight_limit and attempt
+				// withdrawing from the origin of this transaction.
+				let fee = T::WeightToFee::convert(gas_weight_limit);
+
+				// Compute and store the effective price per unit of gas.
+				let gas_price = {
+					let divisor = gas_weight_limit.unique_saturated_into();
+					fee.checked_div(&divisor).unwrap_or(1.into())
+				};
+
+				//
+				// The place where we set GasPrice for the execution of this transaction.
+				//
+				<GasPrice<T>>::put(gas_price);
+
+				let imbalance = match T::Currency::withdraw(
```

Comment:
You call this from `SignedExtension::validate` and `SignedExtension::pre_dispatch`? Aren't you charging the `who` double by doing this?

Reply:
Whatever we do in `validate()` is not persistent. The state is dropped thereafter.

Reply:
But changes in `pre_dispatch` are persisted, right?